File: 04/gpt-5.1/go/main.go
package main

import (
	"fmt"
	"os"
	"strings"
)

func main() {
	data, _ := os.ReadFile("input.txt")
	lines := parseLines(string(data))

	// Part 1
	part1 := countAccessibleRolls(lines)
	fmt.Printf("Day 04 Part 1: %d\n", part1)

	// Part 2
	part2 := countTotalRemoved(lines)
	fmt.Printf("Day 04 Part 2: %d\n", part2)
}

func parseLines(input string) []string {
	raw := strings.Split(strings.ReplaceAll(input, "\r\n", "\n"), "\n")
	lines := make([]string, 0, len(raw))
	for _, line := range raw {
		line = strings.TrimSpace(line)
		if line != "" {
			lines = append(lines, line)
		}
	}
	return lines
}

func countAccessibleRolls(grid []string) int64 {
	height := len(grid)
	if height == 0 {
		return 0
	}
	width := len(grid[0])
	var accessible int64

	dRows := []int{-1, -1, -1, 0, 0, 1, 1, 1}
	dCols := []int{-1, 0, 1, -1, 1, -1, 0, 1}

	for r := 0; r < height; r++ {
		row := grid[r]
		for c := 0; c < width; c++ {
			if row[c] != '@' {
				continue
			}

			neighbors := 0
			for i := 0; i < len(dRows); i++ {
				nr := r + dRows[i]
				nc := c + dCols[i]
				if nr < 0 || nr >= height || nc < 0 || nc >= width {
					continue
				}
				if grid[nr][nc] == '@' {
					neighbors++
				}
			}

			if neighbors < 4 {
				accessible++
			}
		}
	}

	return accessible
}

func countTotalRemoved(grid []string) int64 {
	height := len(grid)
	if height == 0 {
		return 0
	}
	width := len(grid[0])

	board := make([][]byte, height)
	for r := 0; r < height; r++ {
		board[r] = []byte(grid[r])
	}

	dRows := []int{-1, -1, -1, 0, 0, 1, 1, 1}
	dCols := []int{-1, 0, 1, -1, 1, -1, 0, 1}

	var totalRemoved int64

	for {
		toRemove := make([][]bool, height)
		for r := 0; r < height; r++ {
			toRemove[r] = make([]bool, width)
		}

		removedThisRound := 0

		for r := 0; r < height; r++ {
			for c := 0; c < width; c++ {
				if board[r][c] != '@' {
					continue
				}

				neighbors := 0
				for i := 0; i < len(dRows); i++ {
					nr := r + dRows[i]
					nc := c + dCols[i]
					if nr < 0 || nr >= height || nc < 0 || nc >= width {
						continue
					}
					if board[nr][nc] == '@' {
						neighbors++
					}
				}

				if neighbors < 4 {
					toRemove[r][c] = true
					removedThisRound++
				}
			}
		}

		if removedThisRound == 0 {
			break
		}

		for r := 0; r < height; r++ {
			for c := 0; c < width; c++ {
				if toRemove[r][c] {
					board[r][c] = '.'
				}
			}
		}

		totalRemoved += int64(removedThisRound)
	}

	return totalRemoved
}

File: 04/gpt-5.1/rust/src/main.rs
use std::fs;

fn main() {
    let input = fs::read_to_string("input.txt")
        .expect("Failed to read input")
        .trim()
        .to_string();

    let lines: Vec<String> = input
        .lines()
        .map(str::trim)
        .filter(|line| !line.is_empty())
        .map(|s| s.to_string())
        .collect();

    // Part 1
    let part1 = count_accessible_rolls(&lines);
    println!("Day 04 Part 1: {}", part1);

    // Part 2
    let part2 = count_total_removed(&lines);
    println!("Day 04 Part 2: {}", part2);
}

fn count_accessible_rolls(grid: &[String]) -> i64 {
    let height = grid.len();
    if height == 0 {
        return 0;
    }
    let width = grid[0].len();
    let mut accessible: i64 = 0;

    let d_rows = [-1, -1, -1, 0, 0, 1, 1, 1];
    let d_cols = [-1, 0, 1, -1, 1, -1, 0, 1];

    for r in 0..height {
        let row = grid[r].as_bytes();
        for c in 0..width {
            if row[c] != b'@' {
                continue;
            }

            let mut neighbors = 0;
            for i in 0..d_rows.len() {
                let nr = r as isize + d_rows[i];
                let nc = c as isize + d_cols[i];

                if nr < 0
                    || nr >= height as isize
                    || nc < 0
                    || nc >= width as isize
                {
                    continue;
                }

                let nr_usize = nr as usize;
                let nc_usize = nc as usize;
                if grid[nr_usize].as_bytes()[nc_usize] == b'@' {
                    neighbors += 1;
                }
            }

            if neighbors < 4 {
                accessible += 1;
            }
        }
    }

    accessible
}

fn count_total_removed(grid: &[String]) -> i64 {
    let height = grid.len();
    if height == 0 {
        return 0;
    }

    let width = grid[0].len();

    let mut board: Vec<Vec<u8>> = grid
        .iter()
        .map(|row| row.as_bytes().to_vec())
        .collect();

    let d_rows: [isize; 8] = [-1, -1, -1, 0, 0, 1, 1, 1];
    let d_cols: [isize; 8] = [-1, 0, 1, -1, 1, -1, 0, 1];

    let mut total_removed: i64 = 0;

    loop {
        let mut to_remove: Vec<(usize, usize)> = Vec::new();

        for r in 0..height {
            for c in 0..width {
                if board[r][c] != b'@' {
                    continue;
                }

                let mut neighbors = 0;
                for i in 0..d_rows.len() {
                    let nr = r as isize + d_rows[i];
                    let nc = c as isize + d_cols[i];

                    if nr < 0
                        || nr >= height as isize
                        || nc < 0
                        || nc >= width as isize
                    {
                        continue;
                    }

                    if board[nr as usize][nc as usize] == b'@' {
                        neighbors += 1;
                    }
                }

                if neighbors < 4 {
                    to_remove.push((r, c));
                }
            }
        }

        if to_remove.is_empty() {
            break;
        }

        for (r, c) in to_remove.iter() {
            board[*r][*c] = b'.';
        }

        total_removed += to_remove.len() as i64;
    }

    total_removed
}
